stan/math/opencl/rev/copy.hpp

```diff
@@ -0,0 +1,150 @@
+#ifndef STAN_MATH_OPENCL_REV_COPY_HPP
+#define STAN_MATH_OPENCL_REV_COPY_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/opencl/opencl_rev.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <stan/math/prim/arr/fun/vec_concat.hpp>
+
+#include <cl.hpp>
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <type_traits>
+
+namespace stan {
+namespace math {
+
+/**
+ * Copies the source Eigen matrix to
+ * the destination matrix that is stored
+ * on the OpenCL device.
+ *
+ * @tparam R Compile time rows of the Eigen matrix
+ * @tparam C Compile time columns of the Eigen matrix
+ * @param src source Eigen matrix
+ * @return matrix_cl with a copy of the data in the source matrix
+ */
+template <typename T, int R, int C, require_var_t<T>...>
+inline matrix_cl<T> to_matrix_cl(const Eigen::Matrix<T, R, C>& src) try {
+  matrix_cl<T> dst(src.rows(), src.cols());
+  if (src.size() == 0) {
+    return dst;
+  }
+  dst.val() = to_matrix_cl(src.val().eval());
+  dst.adj() = to_matrix_cl(src.adj().eval());
+  return dst;
+} catch (const cl::Error& e) {
+  check_opencl_error("copy var Eigen->(OpenCL)", e);
+  matrix_cl<T> dst(src.rows(), src.cols());
```

Comment:
Ah yes sorry, so this will never actually return since `check_opencl_error()` throws a terminating error. The empty var creation is just so that we don't get a compiler warning about different return types